src/java.base/share/classes/java/util/regex/Matcher.java

```diff
@@ -1372,7 +1378,7 @@ public void forEachRemaining(Consumer<? super MatchResult> action) {
                     return;
 
                 // Capture the input sequence as a string on first find
-                textAsString = text.toString();
+                textAsString = captureText();
```

Comment:
Is this correct? The `find()` on line 1371 will set `first` and `last` and `captureText` uses that to narrow down `textAsString` to that range.. so the next `find()` will then find a new `first` and `last` pair but use the text captured after the first `find`?

Reply:
Yeah this doesn't seem right. To be fair the original code is quite odd. It converts the text to a String once and continues searching over the text thereafter, but stores the String (textAsString) into the returned MatchResults. Note that the text can be mutable (StringBuilder or CharBuffer). It seems like there could be mismatches between `text` and `textAsString`.

I'd think it would be better to search over text consistently, and then capture a subSequence of text from the state of the Matcher at the last moment before creating the MatchResult. As it stands, toMatchResult(String) is confusing because it gets some of its state from the Matcher and some from its argument.

Reply:
It is incorrect, indeed.
Corrected, with additional tests exercising `results()`.